File: src/economy/storageAcceptancePolicy.ts
import type { BuildingState } from '../resources/types.ts';

export const STORAGE_COMMODITY_CODES = {
  firewood: 0,
  timber: 3,
  ale: 6,
  honey: 8,
  wine: 9,
  stone: 10,
  dressedStone: 76,
  polearms: 11,
  ironwork: 12,
  wool: 13,
  cloth: 14,
  barley: 16,
  flax: 18,
  iron: 19,
  clay: 20,
  salt: 21,
  charcoal: 22,
  pottery: 23,
  remedies: 25,
  meat: 28,
  fish: 29,
  berries: 30,
  mushrooms: 31,
  milk: 32,
  apples: 33,
  pears: 4,
  cherries: 34,
  aronia: 5,
  rosehips: 27,
  cabbage: 38,
  carrots: 50,
  beetroot: 53,
  eggs: 36,
  grapes: 37,
  curedMeat: 39,
  smokedFish: 40,
  cheese: 41,
  ryeSheaves: 42,
  oatSheaves: 43,
  barleySheaves: 44,
  maslinSheaves: 45,
  ryeGrain: 46,
  oatGrain: 47,
  maslinGrain: 48,
  ryeFlour: 49,
  maslinFlour: 51,
  ryeBread: 52,
  maslinBread: 54,
  cider: 55,
  mead: 56,
  hides: 58,
  leather: 59,
  shoes: 60,
  jam: 61,
  wax: 64,
  candles: 65,
  pelts: 66,
  yarn: 67,
  linen: 68,
  sidearms: 69,
  shields: 70,
  bows: 71,
  crossbows: 72,
  paddedArmor: 73,
  mailArmor: 74,
  ammunition: 75,
} as const;

export type StorageCommodity = keyof typeof STORAGE_COMMODITY_CODES;

export const STORAGE_COMMODITY_LABELS: Record<StorageCommodity, string> = {
  firewood: 'Firewood',
  timber: 'Timber',
  ale: 'Ale',
  honey: 'Honey',
  wine: 'Wine',
  stone: 'Stone',
  dressedStone: 'Dressed stone',
  wool: 'Wool',
  cloth: 'Clothing',
  barley: 'Threshed barley',
  flax: 'Flax',
  iron: 'Iron ore',
  clay: 'Clay',
  salt: 'Salt',
  charcoal: 'Charcoal',
  pottery: 'Pottery',
  remedies: 'Remedies',
  meat: 'Fresh meat',
  fish: 'Fresh fish',
  berries: 'Raspberries',
  mushrooms: 'Mushrooms',
  milk: 'Milk',
  apples: 'Apples',
  pears: 'Pears',
  cherries: 'Cherries',
  aronia: 'Aronia berries',
  rosehips: 'Rosehips',
  cabbage: 'Cabbage',
  carrots: 'Carrots',
  beetroot: 'Beetroot',
  eggs: 'Eggs',
  grapes: 'Grapes',
  curedMeat: 'Cured meat',
  smokedFish: 'Smoked fish',
  cheese: 'Cheese',
  ryeSheaves: 'Rye sheaves',
  oatSheaves: 'Oat sheaves',
  barleySheaves: 'Barley sheaves',
  maslinSheaves: 'Maslin sheaves',
  ryeGrain: 'Rye grain',
  oatGrain: 'Oat grain',
  maslinGrain: 'Maslin grain',
  ryeFlour: 'Rye flour',
  maslinFlour: 'Maslin flour',
  ryeBread: 'Rye bread',
  maslinBread: 'Maslin bread',
  cider: 'Cider',
  mead: 'Mead',
  hides: 'Untanned hides',
  leather: 'Leather',
  shoes: 'Shoes',
  jam: 'Jam',
  wax: 'Beeswax',
  candles: 'Candles',
  pelts: 'Wild-game pelts',
  yarn: 'Yarn',
  linen: 'Linen',
  polearms: 'Polearms',
  ironwork: 'Ironwork',
  sidearms: 'Sidearms',
  shields: 'Shields',
  bows: 'Bows',
  crossbows: 'Crossbows',
  paddedArmor: 'Padded armor',
  mailArmor: 'Mail armor',
  ammunition: 'Ammunition',
};

export const STOREHOUSE_STORAGE_GROUPS = [
  { label: 'Building materials', commodities: ['timber', 'stone', 'dressedStone'] },
  { label: 'Fuel and minerals', commodities: ['firewood', 'charcoal', 'iron', 'clay', 'salt'] },
  { label: 'Textile materials', commodities: ['wool', 'yarn', 'linen'] },
  { label: 'Market wares', commodities: ['cloth', 'pelts', 'hides', 'leather', 'shoes', 'pottery', 'remedies', 'wax', 'candles'] },
  { label: 'Military stores', commodities: ['ironwork', 'polearms', 'sidearms', 'shields', 'bows', 'crossbows', 'paddedArmor', 'mailArmor', 'ammunition'] },
] as const satisfies ReadonlyArray<{
  label: string;
  commodities: readonly StorageCommodity[];
}>;

export const GRANARY_STORAGE_GROUPS = [
  {
    label: 'Fresh provisions',
    commodities: [
      'meat', 'fish', 'berries', 'mushrooms', 'milk', 'apples', 'pears',
      'cherries', 'aronia', 'rosehips', 'cabbage', 'carrots',
      'beetroot', 'eggs', 'grapes',
    ],
  },
  { label: 'Savory preserves', commodities: ['curedMeat', 'smokedFish', 'cheese'] },
  { label: 'Sweet preserves', commodities: ['honey', 'jam'] },
  { label: 'Drinks', commodities: ['ale', 'cider', 'mead', 'wine'] },
  {
    label: 'Harvest and grain',
    commodities: [
      'ryeSheaves', 'oatSheaves', 'barleySheaves', 'maslinSheaves',
      'ryeGrain', 'oatGrain', 'barley', 'maslinGrain', 'flax',
    ],
  },
  {
    label: 'Milled and baked',
    commodities: ['ryeFlour', 'maslinFlour', 'ryeBread', 'maslinBread'],
  },
] as const satisfies ReadonlyArray<{
  label: string;
  commodities: readonly StorageCommodity[];
}>;

export const STOREHOUSE_STORAGE_COMMODITIES = STOREHOUSE_STORAGE_GROUPS
  .flatMap((group) => [...group.commodities]);
export const GRANARY_STORAGE_COMMODITIES = GRANARY_STORAGE_GROUPS
  .flatMap((group) => [...group.commodities]);

const GRANARY_LEGACY_FRESH = new Set<StorageCommodity>([
  'oatGrain', 'ryeBread', 'maslinBread', 'meat', 'fish', 'berries',
  'mushrooms', 'milk', 'apples', 'pears', 'cherries', 'aronia', 'rosehips',
  'cabbage', 'carrots', 'beetroot', 'eggs', 'grapes',
  'curedMeat', 'smokedFish', 'cheese', 'jam',
]);

function masksAccept(
  lowMask: string | undefined,
  highMask: string | undefined,
  commodity: StorageCommodity,
): boolean {
  const code = STORAGE_COMMODITY_CODES[commodity];
  const high = code >= 64;
  const mask = high ? highMask : lowMask;
  if (mask == null) return true;
  try {
    const bit = BigInt(high ? code - 64 : code);
    return (BigInt(mask) & (1n << bit)) !== 0n;
  } catch {
    return true;
  }
}

export function storageAcceptsCommodity(
  building: Pick<
    BuildingState,
    | 'kind'
    | 'storageAcceptanceMask'
    | 'storageAcceptanceMaskHigh'
    | 'storehouseAcceptsTimber'
    | 'storehouseAcceptsStone'
    | 'storehouseAcceptsFirewood'
    | 'storehouseAcceptsCharcoal'
    | 'storehouseAcceptsIron'
    | 'storehouseAcceptsClay'
    | 'storehouseAcceptsSalt'
    | 'granaryAcceptsFreshFood'
  >,
  commodity: StorageCommodity,
): boolean {
  if (!masksAccept(
    building.storageAcceptanceMask,
    building.storageAcceptanceMaskHigh,
    commodity,
  )) return false;
  if (building.kind === 'granary') {
    return !GRANARY_LEGACY_FRESH.has(commodity) || building.granaryAcceptsFreshFood !== false;
  }
  if (building.kind !== 'village_storehouse') return true;
  switch (commodity) {
    case 'timber': return building.storehouseAcceptsTimber !== false;
    case 'stone': return building.storehouseAcceptsStone !== false;
    case 'firewood': return building.storehouseAcceptsFirewood !== false;
    case 'charcoal': return building.storehouseAcceptsCharcoal !== false;
    case 'iron': return building.storehouseAcceptsIron !== false;
    case 'clay': return building.storehouseAcceptsClay !== false;
    case 'salt': return building.storehouseAcceptsSalt !== false;
    default: return true;
  }
}

export function storageCommodityLabel(commodity: StorageCommodity): string {
  return STORAGE_COMMODITY_LABELS[commodity];
}

export function isStorageCommodity(value: string | undefined): value is StorageCommodity {
  return value != null && value in STORAGE_COMMODITY_CODES;
}

export function renderStorageAcceptanceControls(
  building: BuildingState,
  groups: ReadonlyArray<{ label: string; commodities: readonly StorageCommodity[] }>,
): string {
  const total = groups.reduce((sum, group) => sum + group.commodities.length, 0);
  const active = groups.reduce(
    (sum, group) => sum + group.commodities.filter(
      (commodity) => storageAcceptsCommodity(building, commodity),
    ).length,
    0,
  );
  return `
    <div class="storage-acceptance" aria-label="Accepted goods">
      <div class="storage-acceptance__heading">
        <div>
          <strong>Accepted goods</strong>
          <span>${active} of ${total} enabled</span>
        </div>
        <div class="storage-acceptance__bulk">
          <button type="button" class="resource-action-button storage-acceptance__bulk-action" data-storage-accept-all="true" ${active === total ? 'disabled' : ''}>Accept all</button>
          <button type="button" class="resource-action-button storage-acceptance__bulk-action" data-storage-accept-all="false" ${active === 0 ? 'disabled' : ''}>Accept none</button>
        </div>
      </div>
      ${groups.map((group) => `
        <section class="storage-acceptance__group" aria-label="${group.label}">
          <span class="storage-acceptance__group-label">${group.label}</span>
          <div class="storage-acceptance__grid">
            ${group.commodities.map((commodity) => {
              const label = storageCommodityLabel(commodity);
              const accepts = storageAcceptsCommodity(building, commodity);
              const state = accepts
                ? `${label}: accepting new deliveries.`
                : `${label}: new deliveries blocked.`;
              return `<button type="button" class="resource-action-button resource-action-button--toggle storage-acceptance__commodity resource-cost__item${accepts ? '' : ' is-blocked'}" data-resource-cost="${commodity}" data-storage-commodity="${commodity}" data-storage-accepts="${accepts}" aria-pressed="${accepts}" aria-label="${state}" title="${state}"><span class="resource-cost__icon" aria-hidden="true"></span><span class="storage-acceptance__commodity-label">${label}</span></button>`;
            }).join('')}
          </div>
        </section>
      `).join('')}
    </div>
  `;
}

File: server/src/storage_acceptance_policy.rs
/// Commodity codes double as storage-acceptance bit positions. Codes 64-127
/// live in the companion high mask.
pub const STOREHOUSE_ACCEPTANCE_MASK: u64 = bit(0) // firewood
    | bit(3) // timber
    | bit(10) // stone
    | bit(11) // polearms
    | bit(12) // ironwork
    | bit(13) // wool
    | bit(14) // cloth
    | bit(19) // iron
    | bit(20) // clay
    | bit(21) // salt
    | bit(22) // charcoal
    | bit(23) // pottery
    | bit(25) // remedies
    | bit(58) // hides
    | bit(59) // leather
    | bit(60); // shoes

pub const STOREHOUSE_ACCEPTANCE_MASK_HIGH: u64 = high_bit(64) // wax
    | high_bit(65) // candles
    | high_bit(66) // pelts
    | high_bit(67) // yarn
    | high_bit(68) // linen
    | high_bit(69) // sidearms
    | high_bit(70) // shields
    | high_bit(71) // bows
    | high_bit(72) // crossbows
    | high_bit(73) // padded armor
    | high_bit(74) // mail armor
    | high_bit(75); // ammunition

pub const GRANARY_ACCEPTANCE_MASK: u64 = bit(6) // ale
    | bit(8) // honey
    | bit(9) // wine
    | bit(16) // barley
    | bit(18) // flax
    | bit(28) // meat
    | bit(29) // fish
    | bit(30) // raspberries (legacy field name: berries)
    | bit(31) // mushrooms
    | bit(32) // milk
    | bit(33) // apples
    | bit(34) // cherries
    | bit(36) // eggs
    | bit(37) // grapes
    | bit(39) // cured meat
    | bit(40) // smoked fish
    | bit(41) // cheese
    | bit(42) // rye sheaves
    | bit(43) // oat sheaves
    | bit(44) // barley sheaves
    | bit(45) // maslin sheaves
    | bit(46) // rye grain
    | bit(47) // oat grain
    | bit(48) // maslin grain
    | bit(49) // rye flour
    | bit(51) // maslin flour
    | bit(52) // rye bread
    | bit(54) // maslin bread
    | bit(4) // pears
    | bit(5) // aronia berries
    | bit(27) // rosehips
    | bit(38) // cabbage
    | bit(50) // carrots
    | bit(53) // beetroot
    | bit(55) // cider
    | bit(56) // mead
    | bit(61); // jam

pub const GRANARY_ACCEPTANCE_MASK_HIGH: u64 = 0;

const fn bit(code: u8) -> u64 {
    1u64 << code
}

const fn high_bit(code: u8) -> u64 {
    1u64 << (code - 64)
}

pub fn storage_mask_accepts(mask: u64, commodity_code: u8) -> bool {
    commodity_code < 64 && mask & bit(commodity_code) != 0
}

pub fn storage_masks_accept(low: u64, high: u64, commodity_code: u8) -> bool {
    if commodity_code < 64 {
        storage_mask_accepts(low, commodity_code)
    } else if commodity_code < 128 {
        high & high_bit(commodity_code) != 0
    } else {
        false
    }
}

pub fn storage_kind_acceptance_masks(kind: &str) -> Option<(u64, u64)> {
    match kind {
        "village_storehouse" => Some((STOREHOUSE_ACCEPTANCE_MASK, STOREHOUSE_ACCEPTANCE_MASK_HIGH)),
        "granary" => Some((GRANARY_ACCEPTANCE_MASK, GRANARY_ACCEPTANCE_MASK_HIGH)),
        _ => None,
    }
}

pub fn storage_kind_supports_commodity(kind: &str, commodity_code: u8) -> bool {
    storage_kind_acceptance_masks(kind)
        .is_some_and(|(low, high)| storage_masks_accept(low, high, commodity_code))
}

pub fn set_storage_mask_commodity(mask: u64, commodity_code: u8, accepts: bool) -> u64 {
    if commodity_code >= 64 {
        return mask;
    }
    if accepts {
        mask | bit(commodity_code)
    } else {
        mask & !bit(commodity_code)
    }
}

pub fn set_storage_masks_commodity(
    low: u64,
    high: u64,
    commodity_code: u8,
    accepts: bool,
) -> (u64, u64) {
    if commodity_code < 64 {
        (
            set_storage_mask_commodity(low, commodity_code, accepts),
            high,
        )
    } else if commodity_code < 128 {
        let bit = high_bit(commodity_code);
        (low, if accepts { high | bit } else { high & !bit })
    } else {
        (low, high)
    }
}

pub fn set_storage_masks_all(low: u64, high: u64, kind: &str, accepts: bool) -> (u64, u64) {
    let Some((relevant_low, relevant_high)) = storage_kind_acceptance_masks(kind) else {
        return (low, high);
    };
    if accepts {
        (low | relevant_low, high | relevant_high)
    } else {
        (low & !relevant_low, high & !relevant_high)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn storage_masks_are_specialized_and_default_open() {
        assert!(storage_kind_supports_commodity("village_storehouse", 22));
        assert!(storage_kind_supports_commodity("village_storehouse", 25));
        assert!(!storage_kind_supports_commodity("village_storehouse", 28));
        assert!(storage_kind_supports_commodity("granary", 28));
        assert!(storage_kind_supports_commodity("granary", 9));
        assert!(!storage_kind_supports_commodity("granary", 22));
        assert!(storage_mask_accepts(u64::MAX, 54));
        assert!(storage_kind_supports_commodity("village_storehouse", 64));
        assert!(storage_kind_supports_commodity("village_storehouse", 65));
        assert!(storage_kind_supports_commodity("village_storehouse", 66));
        assert!(storage_kind_supports_commodity("village_storehouse", 67));
        assert!(storage_kind_supports_commodity("village_storehouse", 68));
        assert!(!storage_kind_supports_commodity("granary", 64));
        assert!(storage_masks_accept(u64::MAX, u64::MAX, 65));
        assert!(storage_masks_accept(u64::MAX, u64::MAX, 66));
        assert!(storage_masks_accept(u64::MAX, u64::MAX, 67));
        assert!(storage_masks_accept(u64::MAX, u64::MAX, 68));
    }

    #[test]
    fn individual_and_bulk_changes_preserve_irrelevant_bits() {
        let without_charcoal = set_storage_mask_commodity(u64::MAX, 22, false);
        assert!(!storage_mask_accepts(without_charcoal, 22));
        assert!(storage_mask_accepts(without_charcoal, 28));

        let (no_storehouse_goods, unchanged_high) =
            set_storage_masks_all(u64::MAX, u64::MAX, "village_storehouse", false);
        assert_eq!(no_storehouse_goods & STOREHOUSE_ACCEPTANCE_MASK, 0);
        assert_eq!(unchanged_high & STOREHOUSE_ACCEPTANCE_MASK_HIGH, 0);
        assert!(storage_mask_accepts(no_storehouse_goods, 28));
        let (restored_storehouse_goods, restored_high) = set_storage_masks_all(
            no_storehouse_goods,
            unchanged_high,
            "village_storehouse",
            true,
        );
        assert_eq!(
            restored_storehouse_goods & STOREHOUSE_ACCEPTANCE_MASK,
            STOREHOUSE_ACCEPTANCE_MASK
        );
        assert_eq!(
            restored_high & STOREHOUSE_ACCEPTANCE_MASK_HIGH,
            STOREHOUSE_ACCEPTANCE_MASK_HIGH
        );

        let (low, high) = set_storage_masks_commodity(u64::MAX, u64::MAX, 64, false);
        assert_eq!(low, u64::MAX);
        assert!(!storage_masks_accept(low, high, 64));
        assert!(storage_masks_accept(low, high, 65));

        let (low, high) = set_storage_masks_all(u64::MAX, u64::MAX, "village_storehouse", false);
        assert_eq!(low & STOREHOUSE_ACCEPTANCE_MASK, 0);
        assert_eq!(high & STOREHOUSE_ACCEPTANCE_MASK_HIGH, 0);
    }
}
